soc/xtensa/intel_adsp/common/mem_window.c

```diff
@@ -46,16 +36,56 @@ __imr int mem_win_init(const struct device *dev)
 	return 0;
 }
 
-#define MEM_WIN_INIT(inst)                                                                         \
-	static const struct mem_win_config mem_win_config_##inst = {                               \
-		.base_addr = DT_INST_REG_ADDR(inst),                                               \
-		.size = WIN_SIZE(inst),                                                            \
-		.offset = WIN_OFFSET(inst),                                                        \
-		.read_only = DT_INST_PROP(inst, read_only),					   \
-		.mem_base = DT_REG_ADDR(DT_INST_PHANDLE(inst, memory)) + WIN_OFFSET(inst),         \
-		.initialize = DT_INST_PROP(inst, initialize),                                      \
-	};                                                                                         \
-	DEVICE_DT_INST_DEFINE(inst, mem_win_init, NULL, NULL, &mem_win_config_##inst, ARCH,        \
-			      CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
-
-DT_INST_FOREACH_STATUS_OKAY(MEM_WIN_INIT)
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(mem_window0), okay)
+static const struct mem_win_config mem_win_config_0 = {
+	.base_addr = DT_REG_ADDR(DT_NODELABEL(mem_window0)),
+	.size = WIN_SIZE(0),
+	.offset = WIN_OFFSET(0),
+	.read_only = DT_PROP(DT_NODELABEL(mem_window0), read_only),
+	.mem_base = DT_REG_ADDR(DT_PHANDLE(DT_NODELABEL(mem_window0), memory)) + WIN_OFFSET(0),
+	.initialize = DT_PROP(DT_NODELABEL(mem_window0), initialize),
+};
+DEVICE_DT_DEFINE(DT_NODELABEL(mem_window0), mem_win_init, NULL, NULL, &mem_win_config_0, ARCH,
+		 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(mem_window1), okay)
+static const struct mem_win_config mem_win_config_1 = {
+	.base_addr = DT_REG_ADDR(DT_NODELABEL(mem_window1)),
+	.size = WIN_SIZE(1),
+	.offset = WIN_OFFSET(1),
+	.read_only = DT_PROP(DT_NODELABEL(mem_window1), read_only),
+	.mem_base = DT_REG_ADDR(DT_PHANDLE(DT_NODELABEL(mem_window1), memory)) + WIN_OFFSET(1),
+	.initialize = DT_PROP(DT_NODELABEL(mem_window1), initialize),
+};
+DEVICE_DT_DEFINE(DT_NODELABEL(mem_window1), mem_win_init, NULL, NULL, &mem_win_config_1, ARCH,
+		 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+#endif
```

Comment:
I think it would lead to more maintainable code if you adopted this approach instead: https://github.com/zephyrproject-rtos/zephyr/blob/1d879b93c59e8f1f48d86db5794600769c86facb/drivers/spi/spi_nrfx_spim.c#L617

Reply:
Sorry, that line was confusing. I'm not talking about using a Kconfig symbol. I'm talking about using an approach that's similar, like this excerpt from the same driver:

```c
#define SPI_NRFX_SPIM_DEFINE(n) ...
#if ...
SPI_NRFX_SPIM_DEFINE(0);
#endif

#if ...
SPI_NRFX_SPIM_DEFINE(1);
#endif

...
```

You can do something similar:

```c
#define WINSTREAM_NODE(n) DT_NODELABEL(mem_window##n)
#define WINSTREAM_DEFINE(n) ... { .base_addr = DT_REG_ADDR(WINSTREAM_NODE(n)), ... } ...
#if DT_NODE_HAS_STATUS(WINSTREAM_NODE(0), okay)
WINSTREAM_DEFINE(0)
#endif
#if DT_NODE_HAS_STATUS(WINSTREAM_NODE(1), okay)
WINSTREAM_DEFINE(1)
#endif
...
```

Reply:
agree, I had to get it working first and get the blocks to a unified state first, now I can do the above.